### ISSUE TITLE
v4.0.0-beta.0: HMR fails when some resource that is in path served by webpack-dev-server returns 404

### ISSUE THREAD

**Issue by kroko** [webpack/webpack-dev-server#2880]
- Operating System: macOS 10.15.7 (19H2)
- Node Version: v14.15.1
- NPM Version: 6.14.9
- webpack Version: 5.9.0
- webpack-dev-server Version: v4.0.0-beta.0 (actually, master branch at <https://github.com/webpack/webpack-dev-server/commit/910133d0a2383823504b3a0c2fd21b5a86e733a4>)
- webpack-cli Version: 4.2.0
- Browser: Chrome 87.0.4280.67

- [x] This is a **bug** (IMHO 😄 )
- [ ] This is a **modification** request

### Code

<https://github.com/kroko/webpacktest-devserver-v4-case01>

See README

```
git clone git@github.com:kroko/webpacktest-devserver-v4-case01.git; \
cd webpacktest-devserver-v4-case01; \
npm install; \
npm start
```

All you have to do is flip between "localhost" and "127.0.0.1" in `index.html` and do some dummy changes in `index.scss`.

### Expected Behavior

Accessing non existing assets (404) that are in the path served by webpack-dev-server should not break HMR (ignore 404s).

![Screenshot 2020-11-30 at 17 37 19](https://user-images.githubusercontent.com/720976/100630117-d0d60380-3332-11eb-945a-6bfef4d26578.png)

### Actual Behavior

Accessing non existing assets (404) that are in the path served by webpack-dev-server does break HMR (ignore 404s).

![Screenshot 2020-11-30 at 19 37 44](https://user-images.githubusercontent.com/720976/100644103-8f9a1f80-3343-11eb-8583-d58daba1a7f0.png)

Expected not to break as it was before v4 (it is hard to pinpoint at which moment/combination of upgrading webpack, webpack-cli, webpack-dev-server (webpack-dev-middleware) it stopped working).

### For Bugs; How can we reproduce the behavior?

```
git clone git@github.com:kroko/webpacktest-devserver-v4-case01.git; \
cd webpacktest-devserver-v4-case01; \
npm install; \
npm start
```

All you have to do is flip between "localhost" and "127.0.0.1" in `index.html` and do some dummy changes in `index.scss`.

#### Explanation

I really do not know which is the best palace to address this - here or [webpack-dev-middleware](https://github.com/webpack/webpack-dev-middleware). Please advise.

Chose here as practically the issue is directly observed when using webpack-dev-server.

Long story short - it is for cases where webpack-dev-sever is run with setting `publicPath`s to publicly accessible URL with FQDN for remote live mockuping with team.

HTML has

```html
<link href="http://localhost:4444/assets/index.css" rel="stylesheet" type="text/css"/>
<script src="http://localhost:4444/assets/index.js"></script>
```

when webpack-dev-server is run then development tier is active and CSS is loaded via style-loader not emitted via MiniCssExtractPlugin and thus there is no `index.css` resulting in 404. Those 404 should silently fail (or there should be option to force ignoring 404s). Please see [assertions in readme](https://github.com/kroko/webpacktest-devserver-v4-case01#extra-notes).

**Comment by alexander-akait** [comment]
You use `static` and `index` for assets, so you redirect traffic on `index.html`, expected

**Comment by alexander-akait** [comment]
There it no problems with `publicPath`, because it doesn't even make sense here

**Comment by alexander-akait** [comment]
Remove this https://github.com/kroko/webpacktest-devserver-v4-case01/blob/main/webpack.front.config.js#L145, you don't need it,  https://github.com/kroko/webpacktest-devserver-v4-case01/blob/main/webpack.front.config.js#L133 enough

**Comment by alexander-akait** [comment]
I will describe in the documentation different between `static`/`dev` (same misleading for `headers`)

**Comment by kroko** [comment]
> You use `static` and `index` for assets, so you redirect traffic on `index.html`, expected

Hmm, if you run example repo accessing `http://localhost:4444/assets/index.css` yields 

`Cannot GET /assets/index.css`

No, redirection. What do you mean?

> There is no problems with `publicPath`, because it doesn't even make sense here

Never said there was one 😄 
And also, in this short example it does not make much difference, but it would if some image/font assets were involved (i.e., it matters what goes in `url()`s within the compiled CSS).

> Remove this https://github.com/kroko/webpacktest-devserver-v4-case01/blob/main/webpack.front.config.js#L145, you don't need it, https://github.com/kroko/webpacktest-devserver-v4-case01/blob/main/webpack.front.config.js#L133 enough

Yeah, you mentioned that already in <https://github.com/webpack/webpack-dev-server/issues/2875>, i just stuck it in both places for now.

**Comment by kroko** [comment]
Well, after playing around for a while I found the culprit. 🥳 

Previously
* when building for webpack-dev-server
* and skipping MiniCssExtractPlugin in loaders chain
* and still having reference to CSS file in HTML
* and having MiniCssExtractPlugin dangling around in plugins array
things worked

Now it doesn't. There is some change introduced during the whole ecosystem update which results in `Loading CSS chunk index failed`.

The solution is taking care that if MiniCssExtractPlugin is not used for some build configuration in loaders chain then make sure not to push it in plugins array

```
// ...

        tierIsDevelopment
          ? {
            loader: 'style-loader',
            options: {}
          }
          : {
            loader: MiniCssExtractPlugin.loader,
            options: {
              publicPath: miniCssExtractPublicPath
            }
          },

// ...

// ----------------
// MiniCssExtractPlugin
if (!tierIsDevelopment) { // <= PREVIOUSLY THIS WAS UNECCESARY, ONE COULD LEAVE MiniCssExtractPlugin IN ARRAY
  config.plugins.push(new MiniCssExtractPlugin({
    filename: '[name].css',
    chunkFilename: '[id].css'
  }));
}

// ...
```

Maybe this is something worth mentioning in the future docs as a footnote, a.la., 

> If [MiniCssExtractPlugin](https://webpack.js.org/plugins/mini-css-extract-plugin/) is not used in loader make sure that it is also skipped in plugins array or there might be unexpected results.

**Comment by alexander-akait** [comment]
There is not changed logic for loading CSS chunks, I think you have the problem in the configuration, I don't have time to look at this right now

**Comment by alexander-akait** [comment]
I think the problem in the `publicPath` option, please avoid rewrite it on each level, I left this option to make it convenient to change public path for dev server, now I see developers starting to use it for evil again.

**Comment by kroko** [comment]
> I think the problem in the publicPath option, please avoid rewrite it on each level

It's just a style of trying to be explicit. My actual webpack config file is an essay (it's a mess), because of very specific demands (using webpack to build stuff for HTML frontends that run on embedded devices, involving auto discovery network grid, there are specifics about DNSes, where assets are located n stuff, so there has to be forced (and different) FQDNs/IPs in different places in the build and flightplan chain).

Anyways, closing now. Thanks!
